### DISCUSSION TITLE
Choose a key other than Alt to access accelerator key?

### DISCUSSION THREAD

**Discussion by Mbucari** [AvaloniaUI/Avalonia#10690] [Q&A] [answered]
[One of our users recently pointed out that macOS uses Alt key combinations for special characters](https://github.com/rmcrackan/Libation/issues/534#issue-comment-box). Our application makes use of many MenuItem accelerator keys that conflict with their workflow. Is there a way to instead use the Mac cmd key for accelerator keys?

**Comment by maxkatz6** [comment]
I am pretty sure hotkeys like "cmd+^" or "⌘+^" would also be allowed. 
But it seems like your problem is in supporting different hotkeys depending on the OS. 
In 11.0 previews you can use `="{OnPlatform 'alt+c', macOS='⌘+c'}"`, or any other version you also can define these in C# code and use bindings or `{x:Static}` extension.

**Reply by Mbucari** [reply]
My question is specifically about [Menu Accelerator Keys](https://docs.avaloniaui.net/docs/controls/menu#accelerator-keys), which I believe are different from hotkeys. With hotkeys, I would have to define that property for every MenuItem instead of just using the underscore method in the header text. Is the alt key hard-coded for accelerators and cannot be changed?

**Reply by maxkatz6** [reply]
Ah, no, I don't think it supports other keys https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Input/AccessKeyHandler.cs#L138

**Comment by Mbucari** [comment, marked as answer]
I've successfully changed the accelerator key by overriding AccessKeyHandler and modifying the KeyEventArgs as below, then registering my extension with the locator like so:

```C#
AvaloniaLocator.CurrentMutable.Bind<IAccessKeyHandler>().ToTransient<AccessKeyHandlerEx>();
```

I've discovered the macOS command KeyModifiers is `KeyModifiers.Meta`. 
**My only remaining question is, what is the `Avalonia.Input.Key` for ⌘?**

```C#
internal class AccessKeyHandlerEx : AccessKeyHandler
{
  protected override void OnPreviewKeyDown(object sender, KeyEventArgs e)
  {
    if (Configuration.IsWindows && e.Key is Key.LeftCtrl or Key.RightCtrl)
    {
      var newArgs = new KeyEventArgs
      {
        Key = Key.LeftAlt,
        Handled = e.Handled,
        KeyModifiers = e.KeyModifiers,
      };
      base.OnPreviewKeyDown(sender, newArgs);
      e.Handled = newArgs.Handled;
    }
  }
  
  protected override void OnPreviewKeyUp(object sender, KeyEventArgs e)
  {
    if (Configuration.IsWindows && e.Key is Key.LeftCtrl or Key.RightCtrl)
    {
      var newArgs = new KeyEventArgs()
      {
        Key = Key.LeftAlt,
        Handled = e.Handled,
        KeyModifiers = e.KeyModifiers,
      };
      base.OnPreviewKeyUp(sender, newArgs);
      e.Handled = newArgs.Handled;
    }
  }
  
  
  protected override void OnKeyDown(object sender, KeyEventArgs e)
  {
    if (Configuration.IsWindows && e.KeyModifiers.HasAllFlags(KeyModifiers.Control))
    {
      var newArgs = new KeyEventArgs
      {
        Key = e.Key,
        Handled = e.Handled,
        KeyModifiers = KeyModifiers.Alt,
      };
      base.OnKeyDown(sender, newArgs);
      e.Handled = newArgs.Handled;
    }
  }
}
```

**Reply by timunie** [reply]
> My only remaining question is, what is the Avalonia.Input.Key for ⌘? 

mh, I have no mac, so I can't test. But can't you just set a breakpoint in your OnKeyDown listener and check what value e.Key has? Alternatively, lookup the [keys-enum](https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Input/Key.cs).

**Reply by Mbucari** [reply]
@timunie I don't own a mac. I'm developing an xplat app blind. I'd really appreciate the help on this one 😂

Edit* I checked the keys enum and none looked like they were the Mac command key. My best guess is LWin and RWin double as the command key on Mac, but Id love some confirmation.

**Reply by Mrxx99** [reply]
Not exact the same access key but for browsers, Ctrl + Opt seems to be common for mac: https://en.m.wikipedia.org/wiki/Access_key

**Reply by Mbucari** [reply]
I was right, mac command key is LWin and RWin

**Reply by timunie** [reply]
We should probably add this to the Key-enum description.

**Reply by timunie** [reply]
See: https://github.com/AvaloniaUI/Avalonia/issues/10889
